src/editor/contents.js

```diff
@@ -256,6 +275,87 @@ export default class Contents {
     })
   }
 
+  replaceNodeWithHTML(nodeKey, html, options = {}) {
+    this.editor.update(() => {
+      const node = $getNodeByKey(nodeKey)
+      if (!node) return
+
+      const selection = $getSelection()
+      let wasSelected = false
+
+      if ($isRangeSelection(selection)) {
+        const selectedNodes = selection.getNodes()
+        wasSelected = selectedNodes.includes(node) || selectedNodes.some(n => n.getParent() === node)
+
+        if (wasSelected) {
+          $setSelection(null)
+        }
+      }
+
+      if (options.attachment) {
+        const attachmentConfig = typeof options.attachment === 'object' ? options.attachment : {}
+        const customNode = new CustomActionTextAttachmentNode({
+          sgid: attachmentConfig.sgid || null,
+          contentType: "text/html",
+          innerHtml: html
+        })
+        node.replace(customNode)
+
+        if (wasSelected) {
+          customNode.selectEnd()
+        }
+      } else {
+        const htmlNodes = $generateNodesFromDOM(this.editor, parseHtml(html))
+        if (htmlNodes.length === 0) return
+
+        const firstReplacementNode = htmlNodes[0]
+        node.replace(firstReplacementNode)
+
+        if (htmlNodes.length > 1) {
+          let previousNode = firstReplacementNode
+          htmlNodes.slice(1).forEach(htmlNode => {
+            previousNode.insertAfter(htmlNode)
+            previousNode = htmlNode
+          })
+        }
```

Comment:
I would consider extracting two methods for the two branches, to keep the original method easier to parse and understand. Also, I think there is some duplication you could get rid of: the node replacement and selection. Something like:

```js
const node = options.attachment ? this.#createCustomAttachmentNodeWithHtml(html) : this.#createHtmlNodeWith(html)
node.replace(customNode)
if (wasSelected) {
  node.selectEnd()
}
```